pandas/_libs/algos_take_helper.pxi.in

```diff
@@ -183,8 +183,15 @@ def take_2d_axis1_{{name}}_{{dest}}(ndarray[{{c_type_in}}, ndim=2] values,
                 {{if c_type_in == "uint8_t" and c_type_out == "object"}}
                 out[i, j] = True if values[i, idx] > 0 else False
                 {{else}}
+                {{if c_type_in == "object"}} # GH46848
+                if values[i][idx] is None:
+                    out[i, j] = None
```

Comment:
any idea why this makes a difference?  this looks deeply weird

Reply:
@jbrockmendel i've posted the arguments that cause `take_2d_axis1_object_object` to segfault in https://github.com/pandas-dev/pandas/issues/46848#issuecomment-1147329737 if that helps.

Reply:
perhaps should explicitly check for NULL for clarity. see https://github.com/pandas-dev/pandas/issues/46848#issuecomment-1147458031

Reply:
@simonjayhawkins The PR you refer to is really old, and the current code does not resemble the code of that PR any more. Furthermore I cannot really tell what the change of that PR would be that would help to fix this problem. I guess you are referring to changes in lib.pyx, but the only changes there are in the function call signature of various `isnullobj` variants and an added assert statement.
As far as the check for NULL is concerned, I have determined that the actual value at that point is `None` and not `NULL`. So it would not have any effect.

Reply:
cc @seberg looks like something weird is happening when we create an ndarray with np.empty_like(object_dtype_array) and then pass it to cython.  thoughts?

Reply:
Hmmmpff, It seems `empty_like` does not fill the object array with `None`, but just ensures `NULL`'ing.  NumPy supports this (mostly?), it is perfectly fine if an array is filled with NULL's, this is the same as if the array was filled with `None`.

It is probably best if you also add support for that in pandas.  That said, it would *also* be better to explicitly fill the array with `None` in NumPy.  (And I do think we should define that as the right thing, even if there are branches where NULL's must be expected a fully initialized array filled with `NULL`s should never happen!)

Reply:
To proof the point:
```
In [8]: arr.tobytes()
Out[8]: b'`\xd1\x91\x00\x00\x00\x00\x00`\xd1\x91\x00\x00\x00\x00\x00`\xd1\x91\x00\x00\x00\x00\x00'

In [9]: np.empty_like(arr).tobytes()
Out[9]: b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
```

Reply:
@seberg How would you effectively detect `NULL` values in a numpy array? I assume that @simonjayhawkins prefers a fix somewhere in the `setitem` method of `DataFrame`, but I am at a loss how to intercept those values and convert them to `None`.

Reply:
Honestly, it seems to me that cython may not know about the possibility of NULLs for object typed buffers (either that way or typed memoryviews).  The generated code doesn't look like it.  So I am not sure what a decent way forward would be.

Reply:
So, a work-around that is relatively quick may be the following:
```
    cdef cnp.intp_t[2] idx
    idx[0] = i
    idx[1] = idx

    cdef void *data = cnp.PyArray_GetPtr(values, idx)
    if data == NULL:
        data = <void *>None
    out[i,j] = <object>data
```

But maybe its nicer to just drop that weird `PyArray_GetPtr`:
```
    cdef void *data = arr.data + x * arr.strides[0] + y * arr.strides[1]
    if data == NULL:
        data = <void *>None
    return <object>data
```
which will be as fast as possible without any micro-optimizations (not quite sure on Cython 3, but maybe it doesn't matter).

I am have planning on plugging this hole in NumPy (https://github.com/numpy/numpy/pull/21817), but I am not sure if it is better there or in Cython.
(IMO, it is clearly right to fix this either in NumPy or in Cython, although if we fix it in NumPy, pandas still needs the work-around since you can rely on a new Cython, but not a new NumPy.)